### ISSUE TITLE
deprecated componentWillReceiveProps

### ISSUE THREAD

**Issue by stropho** [parsable/react-truncate-markup#25]
This does **NOT** need an immediate attention, but ..
Earlier in some minor version of react@16, `componentWillReceiveProps` and a few other methods were flagged as `deprecated`. The intention is to remove them completely in react@17.

In this package, both `componentWillReceiveProps` and `componentDidUpdate` do a lot of stuff, and I have no idea how complex it could be (if actually possible) to replace it by the new lifecycle methods. Btw. there is a polyfill for the new methods in older versions of react https://github.com/reactjs/react-lifecycles-compat

**Is there any plan to use the new lifecycle methods ?**

As I said, this does not need to solved right now. However, it would be nice to have some idea about what's going to happen with this package once the release of react@17 is happening.

**Comment by patrik-piskay** [comment]
`componentDidUpdate` will be fine even in the future, but you are correct we need to think how to migrate `componentWillReceiveProps` - the easiest short term solution would be to rename it to `UNSAFE_componentWillReceiveProps` once cWRP is deprecated, but that will not work with the future React's async mode.

I like cWRP though, because we have a nice separation between prop updates (content that may need to be truncated, cWRP) and internal state updates (needed for the truncation, cDU)

**Comment by stropho** [comment]
The way you describe cWRP sounds like it could be replaced with `getDerivedStateFromProps` . It is a static function, so there is no access to `this`. That could make the migration more complicated. I might look into it some day. As it is not a pressing matter now, it might not be anytime soon :wink: 

Feel free to close this issue, or keep it as a reminder .. up to you!

**Comment by bitjoo** [comment]
Hello everyone. Thank you for this nice little component.
What is the status here? Is something planned in the future to get around ` componentWillReceiveProps`?

**Comment by patrik-piskay** [comment]
Yes @bitjoo, thanks for reminding me, I'll plan to take a look soon!

**Comment by patrik-piskay** [comment]
`UNSAFE_componentWillReceiveProps` released in `v4.1.0`

**Comment by stropho** [comment]
So I did some digging, unfortunately not found a clean solution, but at least I can share what I've found.

The `TruncateMarkup` component works in more or less those phases:
1. receive an external update, such as in `UNSAFE_componentWillReceiveProps` or from `resizeObserver`
2. start the truncation process by calling at least 1 `setState`. More `setState` + `componentDidUpdate` could be triggered in order to find the max. content that fits to the container.
3. idle - go back to point 1. with a new update

As **sidenote**, the component handles a case when receiving new props and at the same time the truncation (phase 2) is in progress. Initiating truncation when the previous one hasn't finished could lead to incorrect result. This make me think about questions where I don't really know the answer:
- can component receive new props while in updating process (`setState` called in `componentDidUpdate`) ?
- is this connected to React async rendering ?
- or maybe it could happen just in case the series of `setState`s happen in more than 1 [browser reflow](https://developers.google.com/speed/docs/insights/browser-reflow) :man_shrugging: 


So IMO, the problem is that we need to distinguish what is the internal `setState` (for truncation, phase 2) **vs** external update (phase 1). `UNSAFE_componentWillReceiveProps` solves it quite well for this purpose. Earlier I thought that `getDerivedStateFromProps` can, with some limitations, do the same job - that is not the truth. [Since react@16.4](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html) `getDerivedStateFromProps` is triggered for all updates including those from `setState` - see http://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/ **THERE IS NO EASY WAY TO FIGURE OUT UPDATE IS GOING TO HAPPEN BECAUSE OF PROPS CHANGE**

The only viable solution I could think of is using new `key` or the one mentioned in the [Alternative 2](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#alternative-2-reset-uncontrolled-component-with-an-instance-method)

For example the library could internally have 2 components
```jsx
class WrapperThatRerendersWithNewProps extends React.PureComponent {
    render() {
      return <TruncateMarkup 
       {...this.props}
       key='new-key-for-every-rerender-that-effectively-remounts-TruncateMarkup'     
      />
  }
}
```

Any be better ideas? There has to be something :grimacing:

**Comment by patrik-piskay** [comment]
Yes, I believe having the library consumers to set `key` to remount is the best option we have. 

I don't think the "updating with new props" case is something we need to support, mainly because I'm not aware of use cases where updating with new props is necessary, except maybe when fetching the data from server. But that should be handled by conditional rendering (rendering RTM only once the data is ready).

For other use cases where the text changes dynamically, it's probably reasonable to expect a new `key` provided by users.

Regarding your proposed `WrapperThatRerendersWithNewProps` type of a component, I think that should also be the kind of abstraction that should be handled by users, if that's what they want to do.

**Comment by mundanelunacy** [comment]
IMHO, best solution would be to refactor to a functional component using useState and useEffect hooks

**Comment by ramosbugs** [comment]
React [Strict Mode](https://reactjs.org/docs/strict-mode.html) complains about this now:
```
Warning: Using UNSAFE_componentWillReceiveProps in strict mode is not recommended and may indicate bugs in your code. See https://reactjs.org/link/unsafe-component-lifecycles for details.

* Move data fetching code or side effects to componentDidUpdate.
* If you're updating state whenever props change, refactor your code to use memoization techniques or move it to static getDerivedStateFromProps. Learn more at: https://reactjs.org/link/derived-state

Please update the following components: TruncateMarkup
```

If I'm reading the docs right, it looks like the main consequence is that the future React concurrent rendering feature won't work with unsafe lifecycles.

**Comment by bonesoul** [comment]
i'm also getting the same warning @ramosbugs , any solutions?

**Comment by elishaterada** [comment]
Any plan to address UNSAFE_componentWillReceiveProps issue? If someone in the community converts the class component to functional, would PR be reviewed and accepted? Or are we better off forking and releasing it under the new npm name? I use this package often so I might take time to help with this.

**Comment by patrik-piskay** [comment]
Absolutely, I've taken a stab at it a while back when the deprecation was
announced but the functional component for RTM didn't work out because of
how the cWRP is used here. I haven't managed to get back to it so if you'd
like to help out, any PR is appreciated!

**Comment by stropho** [comment]
I am not the maintainer but I've contributed to this repo a few times. IMHO if the new functional component works the same - accepting same props, displays the same result and does not introduce performance issues, I don't see why it shouldn't be merged.

**Comment by bonesoul** [comment]
any changes we'll get a fix soon?

**Comment by hc0503** [comment]
this issue is yet happened.

**Comment by h0gar** [comment]
For what it's worth, I have asked Claude to rewrite it as a functional component.

```
import React, { useState, useRef, useEffect, useMemo } from 'react';
import PropTypes from 'prop-types';
import memoizeOne from 'memoize-one';
import getLineHeight from 'line-height';
import ResizeObserver from 'resize-observer-polyfill';
import TOKENIZE_POLICY from './tokenize-rules';
import { Atom, isAtomComponent, ATOM_STRING_ID } from './atom';

const SPLIT = {
  LEFT: true,
  RIGHT: false
};

const toString = (node, string = '') => {
  if (!node) {
    return string;
  } else if (typeof node === 'string') {
    return string + node;
  } else if (isAtomComponent(node)) {
    return string + ATOM_STRING_ID;
  }
  
  const children = Array.isArray(node) ? node : node.props.children || '';
  
  return string + React.Children.map(children, child => 
    toString(child)
  ).join('');
};

const cloneWithChildren = (node, children, isRootEl, level) => {
  const getDisplayStyle = () => {
    if (isRootEl) {
      return {
        // root element cannot be an inline element because of the line calculation
        display: (node.props.style || {}).display || 'block'
      };
    } else if (level === 2) {
      return {
        // level 2 elements (direct children of the root element) need to be inline because of the ellipsis.
        // if level 2 element was a block element, ellipsis would get rendered on a new line, breaking the max number of lines
        display: (node.props.style || {}).display || 'inline-block'
      };
    } else return {};
  };

  return {
    ...node,
    props: {
      ...node.props,
      style: {
        ...node.props.style,
        ...getDisplayStyle()
      },
      children
    }
  };
};

const validateTree = (node) => {
  if (node == null || ['string', 'number'].includes(typeof node) || isAtomComponent(node)) {
    return true;
  } else if (typeof node.type === 'function') {
    if (process.env.NODE_ENV !== 'production') {
      /* eslint-disable no-console */
      console.error(`ReactTruncateMarkup tried to render <${node.type.name} />, but truncating React components is not supported, the full content is rendered instead. Only DOM elements are supported. Alternatively, you can take advantage of the <TruncateMarkup.Atom /> component (see more in the docs https://github.com/patrik-piskay/react-truncate-markup/blob/master/README.md#truncatemarkupatom-).`);
      /* eslint-enable */
    }
    return false;
  }

  if (node.props && node.props.children) {
    return React.Children.toArray(node.props.children).reduce(
      (isValid, child) => isValid && validateTree(child),
      true
    );
  }

  return true;
};

const TruncateMarkup = ({
  children,
  lines = 1,
  ellipsis = '...',
  lineHeight: propLineHeight = '',
  onTruncate = () => {},
  tokenize = 'characters'
}) => {
  const elRef = useRef(null);
  const resizeObserverRef = useRef(null);
  const [text, setText] = useState(null);
  
  // Refs instead of state to avoid re-renders
  const lineHeightRef = useRef(null);
  const shouldTruncateRef = useRef(true);
  const wasLastCharTestedRef = useRef(false);
  const endFoundRef = useRef(false);
  const latestThatFitsRef = useRef(null);
  const onTruncateCalledRef = useRef(false);
  const splitDirectionSeqRef = useRef([]);
  
  const toStringMemo = useMemo(() => memoizeOne(toString), []);
  const validateTreeMemo = useMemo(() => memoizeOne(validateTree), []);
  
  const policy = TOKENIZE_POLICY[tokenize] || TOKENIZE_POLICY.characters;
  
  const isValid = useMemo(() => validateTreeMemo(children), [children, validateTreeMemo]);

  const childrenElementWithRef = useMemo(() => {
    const child = React.Children.only(children);
    return React.cloneElement(child, {
      ref: elRef,
      style: {
        wordWrap: 'break-word',
        ...child.props.style
      }
    });
  }, [children]);

  const fits = () => {
    if (!elRef.current) return true;
    
    const { height } = elRef.current.getBoundingClientRect();
    const computedLines = Math.round(height / parseFloat(lineHeightRef.current));
    
    return lines >= computedLines;
  };

  const handleResize = (el, prevResizeObserver) => {
    // clean up previous observer
    if (prevResizeObserver) {
      prevResizeObserver.disconnect();
    }

    // unmounting or just unsetting the element to be replaced with a new one later
    if (!el) return null;

    /* Wrapper element resize handling */
    let initialRender = true;
    const resizeCallback = () => {
      if (initialRender) {
        // ResizeObserver cb is called on initial render too so we are skipping here
        initialRender = false;
      } else {
        // wrapper element has been resized, recalculating with the original text
        shouldTruncateRef.current = false;
        latestThatFitsRef.current = null;

        setText(childrenElementWithRef);
        
        setTimeout(() => {
          shouldTruncateRef.current = true;
          onTruncateCalledRef.current = false;
          truncate();
        }, 0);
      }
    };

    const resizeObserver = prevResizeObserver || new ResizeObserver(resizeCallback);
    resizeObserver.observe(el);

    return resizeObserver;
  };

  const splitString = (string, splitDirections = [], level) => {
    if (!splitDirections.length) {
      return string;
    }

    if (splitDirections.length && policy.isAtomic(string)) {
      // allow for an extra render test with the current character included
      // in most cases this variation was already tested, but some edge cases require this check
      if (!wasLastCharTestedRef.current) {
        wasLastCharTestedRef.current = true;
      } else {
        // we are trying to split further but we have nowhere to go now
        // that means we've already found the max subtree that fits the container
        endFoundRef.current = true;
      }

      return string;
    }

    if (policy.tokenizeString) {
      const wordsArray = splitArray(policy.tokenizeString(string), splitDirections, level);
      // in order to preserve the input structure
      return wordsArray.join('');
    }

    const [splitDirection, ...restSplitDirections] = splitDirections;

    const pivotIndex = Math.ceil(string.length / 2);
    const beforeString = string.substring(0, pivotIndex);

    if (splitDirection === SPLIT.LEFT) {
      return splitString(beforeString, restSplitDirections, level);
    }
    
    const afterString = string.substring(pivotIndex);
    return beforeString + splitString(afterString, restSplitDirections, level);
  };

  const splitArray = (array, splitDirections = [], level) => {
    if (!splitDirections.length) {
      return array;
    }

    if (array.length === 1) {
      return [split(array[0], splitDirections, false, level)];
    }

    const [splitDirection, ...restSplitDirections] = splitDirections;

    const pivotIndex = Math.ceil(array.length / 2);
    const beforeArray = array.slice(0, pivotIndex);

    if (splitDirection === SPLIT.LEFT) {
      return splitArray(beforeArray, restSplitDirections, level);
    }
    
    const afterArray = array.slice(pivotIndex);
    return beforeArray.concat(splitArray(afterArray, restSplitDirections, level));
  };

  const split = (node, splitDirections, isRoot = false, level = 1) => {
    if (!node || isAtomComponent(node)) {
      endFoundRef.current = true;
      return node;
    } else if (typeof node === 'string') {
      return splitString(node, splitDirections, level);
    } else if (Array.isArray(node)) {
      return splitArray(node, splitDirections, level);
    }

    const newChildren = split(
      node.props.children, 
      splitDirections,
      false, 
      level + 1
    );

    return cloneWithChildren(node, newChildren, isRoot, level);
  };

  const tryToFit = (rootEl, splitDirections) => {
    if (!rootEl.props.children) {
      // no markup in container
      return;
    }

    const newRootEl = split(rootEl, splitDirections, true);

    const ellipsisContent = typeof ellipsis === 'function' ? ellipsis(newRootEl) : ellipsis;
    const ellipsisElement = typeof ellipsisContent === 'object' 
      ? React.cloneElement(ellipsisContent, { key: 'ellipsis' }) 
      : ellipsisContent;

    const newChildren = newRootEl.props.children;
    const newChildrenWithEllipsis = [].concat(newChildren, ellipsisElement);

    // Check if we need to show ellipsis by comparing the new text with original
    const shouldRenderEllipsis = toString(newChildren) !== toStringMemo(children);

    setText({
      ...newRootEl,
      props: {
        ...newRootEl.props,
        children: shouldRenderEllipsis ? newChildrenWithEllipsis : newChildren
      }
    });
  };

  const truncateOriginalText = () => {
    endFoundRef.current = false;
    splitDirectionSeqRef.current = [SPLIT.LEFT];
    wasLastCharTestedRef.current = false;

    tryToFit(childrenElementWithRef, splitDirectionSeqRef.current);
  };

  const truncate = () => {
    if (fits()) {
      // the whole text fits on the first try, no need to do anything else
      shouldTruncateRef.current = false;
      onTruncate(false); // wasTruncated = false
      return;
    }

    truncateOriginalText();
  };

  // Initial setup
  useEffect(() => {
    if (!isValid) {
      return;
    }

    setText(childrenElementWithRef);
    
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // Set up lineHeight and initial truncation
  useEffect(() => {
    if (!isValid || !elRef.current) {
      return;
    }

    // get the computed line-height of the parent element
    lineHeightRef.current = propLineHeight || getLineHeight(elRef.current);
    truncate();
    
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isValid, propLineHeight]);

  // Handle resize observer
  useEffect(() => {
    if (elRef.current) {
      resizeObserverRef.current = handleResize(elRef.current, resizeObserverRef.current);
    }
    
    return () => {
      if (resizeObserverRef.current) {
        resizeObserverRef.current.disconnect();
      }
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [elRef.current]);

  // Update when children changes
  useEffect(() => {
    shouldTruncateRef.current = false;
    latestThatFitsRef.current = null;
    
    setText(childrenElementWithRef);
    
    if (!isValid) {
      return;
    }
    
    lineHeightRef.current = propLineHeight || (elRef.current ? getLineHeight(elRef.current) : null);
    shouldTruncateRef.current = true;
    onTruncateCalledRef.current = false;
    truncate();
    
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [children, propLineHeight, lines, ellipsis, tokenize]);

  // Check if content fits after each render
  useEffect(() => {
    if (!shouldTruncateRef.current || !isValid) {
      return;
    }

    if (endFoundRef.current) {
      // we've found the end where we cannot split the text further
      // that means we've already found the max subtree that fits the container
      if (latestThatFitsRef.current !== null && text !== latestThatFitsRef.current) {
        setText(latestThatFitsRef.current);
        return;
      }

      if (!onTruncateCalledRef.current) {
        onTruncateCalledRef.current = true;
        onTruncate(true); // wasTruncated = true
      }
      return;
    }

    if (splitDirectionSeqRef.current.length) {
      if (fits()) {
        latestThatFitsRef.current = text;
        // we've found a subtree that fits the container
        // but we need to check if we didn't cut too much of it off
        // so we are changing the last splitting decision from splitting and going left
        // to splitting and going right
        splitDirectionSeqRef.current.splice(
          splitDirectionSeqRef.current.length - 1, 
          1, 
          SPLIT.RIGHT, 
          SPLIT.LEFT
        );
      } else {
        splitDirectionSeqRef.current.push(SPLIT.LEFT);
      }

      tryToFit(childrenElementWithRef, splitDirectionSeqRef.current);
    }
    
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [text]);

  if (!text) {
    return null;
  }

  return text;
};

TruncateMarkup.Atom = Atom;

TruncateMarkup.propTypes = process.env.NODE_ENV !== "production" ? {
  children: PropTypes.element.isRequired,
  lines: PropTypes.number,
  ellipsis: PropTypes.oneOfType([PropTypes.element, PropTypes.string, PropTypes.func]),
  lineHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onTruncate: PropTypes.func,
  // eslint-disable-next-line
  onAfterTruncate: function(props, propName, componentName) {
    if (props[propName]) {
      return new Error(`${componentName}: Setting \`onAfterTruncate\` prop is deprecated, use \`onTruncate\` instead.`);
    }
  },
  tokenize: function(props, propName, componentName) {
    const tokenizeValue = props[propName];

    if (typeof tokenizeValue !== 'undefined') {
      if (!TOKENIZE_POLICY[tokenizeValue]) {
        return new Error(`${componentName}: Unknown option for prop 'tokenize': '${tokenizeValue}'. Option 'characters' will be used instead.`);
      }
    }
  }
} : {};

export default TruncateMarkup;
```
